pkg/util/workqueue/delaying_queue.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workqueue
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// DelayingInterface is an Interface that can Add an item at a later time.  This makes it easier to
+// requeue items after failures without ending up in a hot-loop.
+type DelayingInterface interface {
+	Interface
+	AddAfter(item interface{}, duration time.Duration)
+}
+
+// NewDelayingQueue constructs a new workqueue with delayed queuing ability
+func NewDelayingQueue() DelayingInterface {
+	return newDelayingQueue(util.RealClock{})
+}
+
+func newDelayingQueue(clock util.Clock) DelayingInterface {
+	ret := &delayingType{
+		Type:          New(),
+		clock:         clock,
+		waitingCond:   sync.NewCond(&sync.Mutex{}),
+		nextCheckTime: clock.Now().Add(60 * time.Second),
+	}
+
+	go ret.waitingLoop()
+	ret.notifyAt(clock.Now().Add(maxWait))
+
+	return ret
+}
+
+type delayingType struct {
```

Comment:
> I know this is internal, and maybe it should all be obvious, but it would be useful to have comments explaining each struct and its members.

Will do.